packages/block-library/src/gallery/gallery-image.js

```diff
@@ -18,11 +19,23 @@ class GalleryImage extends Component {
 	constructor() {
 		super( ...arguments );
 
+		this.onBlur = this.onBlur.bind( this );
+		this.onFocus = this.onFocus.bind( this );
 		this.onSelectImage = this.onSelectImage.bind( this );
 		this.onSelectCaption = this.onSelectCaption.bind( this );
 		this.onRemoveImage = this.onRemoveImage.bind( this );
 		this.bindContainer = this.bindContainer.bind( this );
 
+		// debouncedOnSelect will be called every time any figure's element
+		// is blurred. Every time a figure's element is focused, it'll be cancelled.
+		//
+		// We use this to detect whether the figure element has lost focus permanently
+		// or the change was internal (a focus transition from image to caption, for example).
+		//
+		// onBlur / onFocus events are quick operations (<5ms apart in my testing),
+		// so 50ms accounts for 10x lagging while feels responsive to the user.
+		this.debouncedOnDeselect = debounce( this.props.onDeselect, 50 );
```

Comment:
I still don't get from this comment why focus on the same element would be called right after blur. In what circumstances does rapid focus & blur happen?

Reply:
Updated the comment to better explain the mechanism: the idea is to call `onDeselect` when focus has transitioned to a different React component (another `GalleryImage`, toolbar, etc.) but not when focus transitions within the DOM elements that form this React component (figure, image, caption, toolbar, etc).

Does it make more sense now?

Reply:
@ellatrix any chance you'd be able to review this and see if the comment makes more sense now?

Reply:
Can't we instead just check that the `window.activeElement` property is not a children of the wrapper (using a ref) instead of this magic trick :)

Reply:
It's just code! Perhaps the comment is more confusing than the code itself? Suggestions welcome :)

I had tried the comparison with the `activeElement`, but the global `focus` handler won't be reliably dispatched, so I resorted to the `focus` events dispatched by the children instead.

Reply:
> I still don't get from this comment why focus on the same element would be called right after blur. In what circumstances does rapid focus & blur happen?

Now that I revisit this thread I think I can explain this better! It's not that rapid focus and blur happen on the same element, but that focus and blur on children element bubble to figure. In the DOM, blur and focus don't bubble, but React Events mechanism normalizes all events to bubble - even blur and focus. So this is how events work in gallery:

- user click image => focus event is fired with image as target, then it bubbles to figure
- user tabs to figcaption =>
  - blur event is fired for image as target, then it bubbles to figure
  - focus event is fired for figcaption, then it bubbles to figure

And the same thing for the GalleryImage buttons (icons to remove or move the image). So, instead of tracking blur events individually, this tracks them in a single place.

Also worth noting that if you check the `activeElement` in a blur event it'll be `body` in Firefox and Chrome (it doesn't give you the next element to focus). So, the only alternative I see would be to listen to focus events at the window level, and check if the `activeElement` is one of the GalleryImage children. This doesn't work because the focus won't bubble in some situations.

Reply:
@nosolosw Could you do something like `event.currentTarget.contains( event.relatedTarget )` in the `onBlur` to detect when focus moved outside a gallery image?

Reply:
That would work and it seems to have [good browser support](https://caniuse.com/#feat=mdn-api_focusevent_relatedtarget). However, the feature is marked as [experimental and is only in a working draft](https://developer.mozilla.org/es/docs/Web/API/FocusEvent/relatedTarget). My gut feeling is that we shouldn't use it just yet, what do you think?